### PR TITLE
feat: Create and update a default user avatar - MEED-3111- Meeds-io/MIPs#83

### PR DESCRIPTION
**Pull request by SaraBoutej**
This PR allows to implement a new default user avatar based on the user’s full name, so if a user has not updated their profile image, a new avatar with a random color and a text with the first letter of its username and surname will be generated automatically and will only be updated if the user changes his profile image of his first or last name.

### CODE REVIEW

File: component/core/src/main/java/org/exoplatform/social/core/image/ImageUtils.java [hunk 1]
```diff
@@ -25,18 +25,22 @@
...
 public class ImageUtils {
   public static final String KEY_SEPARATOR           = "_";
+
   public static final String KEY_DIMENSION_SEPARATOR = "x";
 
   public static final String GIF_EXTENDSION          = "gif";
+
+  private static final int   DEFAULT_AVATAR_WIDTH    = 350;
+
+  private static final int   DEFAULT_AVATAR_HEIGHT   = 350;
```
Comment (boubaker) [inline component/core/src/main/java/org/exoplatform/social/core/image/ImageUtils.java:43]: Formatting, missing new line

File: component/core/src/main/java/org/exoplatform/social/core/image/ImageUtils.java [hunk 2]
```diff
@@ -128,27 +132,75 @@ public static AvatarAttachment createResizedAvatarAttachment(InputStream imageSt
       }
 
       // Create temp file to store resized image to put to avatar attachment
-      File tmp = File.createTempFile("RESIZED", null);
+      ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
```
Comment (boubaker) [inline component/core/src/main/java/org/exoplatform/social/core/image/ImageUtils.java]: may be it's better to not change here, sorry I wasn't clear may be about this change which finally should be applied on Default Avatars only to not have to bring huge avatar files in memory, in contrary to default avatars which is less heavy.

File: component/core/src/main/java/org/exoplatform/social/core/image/ImageUtils.java [hunk 3]
```diff
@@ -128,27 +132,75 @@ public static AvatarAttachment createResizedAvatarAttachment(InputStream imageSt
...
+
+      int x = (DEFAULT_AVATAR_WIDTH - fm.stringWidth(fullNameAbbreviation)) / 2;
+      int y = (fm.getAscent() + (DEFAULT_AVATAR_HEIGHT - (fm.getAscent() + fm.getDescent())) / 2);
+
+      graphics.drawString(fullNameAbbreviation, x, y);
+      graphics.drawImage(image, 0, 0, DEFAULT_AVATAR_WIDTH, DEFAULT_AVATAR_HEIGHT, null);
+      graphics.dispose();
+
+      ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+    try {
```
Comment (boubaker) [inline component/core/src/main/java/org/exoplatform/social/core/image/ImageUtils.java:193]: formatting, not on the same line.

File: component/core/src/main/java/org/exoplatform/social/core/jpa/storage/RDBMSIdentityStorageImpl.java [hunk 4]
```diff
@@ -287,6 +290,15 @@ private void mapToProfileEntity(Profile profile, IdentityEntity identityEntity)
 
       } else if (!Profile.EXPERIENCES_SKILLS.equals(profileProperty.getKey())) {
         Object val = profileProperty.getValue();
+        if ((Profile.FIRST_NAME.equalsIgnoreCase(profileProperty.getKey())
+            || Profile.LAST_NAME.equalsIgnoreCase(profileProperty.getKey()))
+            && !profileProperty.getValue().equals(entityProperties.get(profileProperty.getKey()))) {
```
Comment (boubaker) [inline component/core/src/main/java/org/exoplatform/social/core/jpa/storage/RDBMSIdentityStorageImpl.java]: Must be sure that avatar exists before calling `getAvatarFile` which will generate a new File if not existing.

```suggestion
            && !profileProperty.getValue().equals(entityProperties.get(profileProperty.getKey()))
            && identityEntity.getAvatarFileId() != null) {
```

File: component/core/src/main/java/org/exoplatform/social/core/jpa/storage/RDBMSIdentityStorageImpl.java [hunk 5]
```diff
@@ -1107,6 +1143,15 @@ public ActivityStorage getActivityStorage() {
     return activityStorage;
   }
 
+  private static String getNameAbbreviation(String name) {
```
Comment (boubaker) [inline component/core/src/main/java/org/exoplatform/social/core/jpa/storage/RDBMSIdentityStorageImpl.java]: To move in the bottom of the file after the `public` methods (Code conventions, public methods should be declared before private ones)
Besides, no need for the `static` modifer

```suggestion
  private String getNameAbbreviation(String name) {
```

File: component/core/src/main/java/org/exoplatform/social/core/service/LinkProvider.java [hunk 6]
```diff
@@ -471,7 +471,6 @@ private static String buildAttachmentUrl(String providerId,
     }
 
     if (lastModifiedDate == null || lastModifiedDate <= 0 || lastModifiedDate == DEFAULT_IMAGES_LAST_MODIFED) {
-      id = DEFAULT_IMAGE_REMOTE_ID;
```
Comment (boubaker) [inline component/core/src/main/java/org/exoplatform/social/core/service/LinkProvider.java]: We must keep this to preserve the current behavior concerning default avatars for other type of identities

File: component/core/src/main/java/org/exoplatform/social/core/storage/cache/SocialStorageCacheService.java [hunk 7]
```diff
@@ -129,7 +132,9 @@ public ExoCache<IdentityFilterKey, IntegerData> getCountIdentitiesCache() {
   public ExoCache<ListIdentitiesKey, ListIdentitiesData> getIdentitiesCache() {
     return identitiesCache;
   }
-
+  public ExoCache<IdentityKey, Long> getProfileAvatarCache() {
+    return profileAvatarCache;
+  }
```
Comment (boubaker) [inline component/core/src/main/java/org/exoplatform/social/core/storage/cache/SocialStorageCacheService.java:137]: Formatting, line break

```suggestion
  }

```

File: component/core/src/test/java/org/exoplatform/social/core/jpa/storage/IdentityStorageTest.java [hunk 8]
```diff
@@ -813,11 +813,12 @@ public void testGetAvatarInputStreamById() throws Exception {
     // within this instruction the profile is created implicitly and it does not have an avatar
     String identityId = identity.getId();
     assertNotNull(identityId);
+
     InputStream stream = identityStorage.getAvatarInputStreamById(identity);
-    assertNull(stream);
+    assertNotNull(stream);
```
Comment (boubaker) [inline component/core/src/test/java/org/exoplatform/social/core/jpa/storage/IdentityStorageTest.java:820]: May be add more tests about Avatar File identifier which shouldn't be null + File name + `isDefaultAvatar` to add in Profile + make sure that avatarLastUpdateTime is updated after renaming an identity of type user.

File: component/core/src/main/java/org/exoplatform/social/core/image/ImageUtils.java (hunk 2 above)
Comment (SaraBoutej) [reply component/core/src/main/java/org/exoplatform/social/core/image/ImageUtils.java]: fixed

File: component/core/src/main/java/org/exoplatform/social/core/image/ImageUtils.java (hunk 1 above)
Comment (SaraBoutej) [reply component/core/src/main/java/org/exoplatform/social/core/image/ImageUtils.java:43]: done

File: component/core/src/main/java/org/exoplatform/social/core/image/ImageUtils.java (hunk 3 above)
Comment (SaraBoutej) [reply component/core/src/main/java/org/exoplatform/social/core/image/ImageUtils.java:193]: done

File: component/core/src/main/java/org/exoplatform/social/core/jpa/storage/RDBMSIdentityStorageImpl.java (hunk 4 above)
Comment (SaraBoutej) [reply component/core/src/main/java/org/exoplatform/social/core/jpa/storage/RDBMSIdentityStorageImpl.java]: done

File: component/core/src/main/java/org/exoplatform/social/core/storage/cache/SocialStorageCacheService.java (hunk 7 above)
Comment (SaraBoutej) [reply component/core/src/main/java/org/exoplatform/social/core/storage/cache/SocialStorageCacheService.java:137]: fixed

File: component/core/src/test/java/org/exoplatform/social/core/jpa/storage/IdentityStorageTest.java (hunk 8 above)
Comment (SaraBoutej) [reply component/core/src/test/java/org/exoplatform/social/core/jpa/storage/IdentityStorageTest.java:820]: done

File: component/service/src/test/java/org/exoplatform/social/rest/impl/users/UserRestResourcesTest.java [hunk 9]
```diff
@@ -535,10 +535,6 @@ public void testGetUserAvatarForAnonymous() throws Exception {
     response = service("GET", getURLResource("users/" + user + "/avatar"), "", null, null);
     assertNotNull(response);
     assertEquals(404, response.getStatus());
-
-    response = service("GET", getURLResource("users/" + LinkProvider.DEFAULT_IMAGE_REMOTE_ID + "/avatar"), "", null, null);
-    assertNotNull(response);
-    assertEquals(200, response.getStatus());
```
Comment (boubaker) [inline component/service/src/test/java/org/exoplatform/social/rest/impl/users/UserRestResourcesTest.java]: Should remain true and we shouldn't modify this.

File: component/service/src/test/java/org/exoplatform/social/rest/impl/users/UserRestResourcesTest.java [hunk 10]
```diff
@@ -558,10 +554,6 @@ public void testGetUserAvatarForAuthentiticatedUser() throws Exception {
     response = service("GET", getURLResource("users/" + user + "/avatar"), "", null, null);
     assertNotNull(response);
     assertEquals(200, response.getStatus());
-
-    response = service("GET", getURLResource("users/" + LinkProvider.DEFAULT_IMAGE_REMOTE_ID + "/avatar"), "", null, null);
-    assertNotNull(response);
-    assertEquals(200, response.getStatus());
```
Comment (boubaker) [inline component/service/src/test/java/org/exoplatform/social/rest/impl/users/UserRestResourcesTest.java]: Should remain true and we shouldn't modify this.

File: component/service/src/test/java/org/exoplatform/social/rest/impl/users/UserRestResourcesTest.java (hunk 9 above)
Comment (SaraBoutej) [reply component/service/src/test/java/org/exoplatform/social/rest/impl/users/UserRestResourcesTest.java]: fixed

File: component/service/src/test/java/org/exoplatform/social/rest/impl/users/UserRestResourcesTest.java (hunk 10 above)
Comment (SaraBoutej) [reply component/service/src/test/java/org/exoplatform/social/rest/impl/users/UserRestResourcesTest.java]: fixed

**Review by boubaker — APPROVED** [review]
